klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
+    pub fn populate_unicodes_to_retain(
+        &mut self,
+        input_gids: &IntSet<GlyphId>,
+        input_unicodes: &IntSet<u32>,
+        font: &FontRef,
+    ) {
+        let charmap = font.charmap();
+        if input_gids.is_empty() && input_unicodes.len() < self.font_num_glyphs {
+            let cap = input_unicodes.len();
+            self.unicode_to_new_gid_list.reserve(cap);
+            self.codepoint_to_glyph.reserve(cap);
+            //TODO: add support for subset accelerator?
+
+            for cp in input_unicodes.iter() {
+                match charmap.map(cp) {
+                    Some(gid) => {
+                        self.codepoint_to_glyph.insert(cp, gid);
+                        self.unicode_to_new_gid_list.push((cp, gid));
+                    }
+                    None => {
+                        continue;
+                    }
+                }
             }
-            None => {
-                continue;
+        } else {
+            //TODO: add support for subset accelerator?
+            let charmap = font.charmap();
+            let cmap_unicodes = charmap.mappings().map(|t| t.0).collect::<IntSet<u32>>();
+            let unicode_gid_map = charmap.mappings().collect::<HashMap<u32, GlyphId>>();
+
+            let mut vec_cap = input_gids.len() + input_unicodes.len();
+            vec_cap = vec_cap.min(cmap_unicodes.len());
+            self.codepoint_to_glyph.reserve(vec_cap);
+            self.unicode_to_new_gid_list.reserve(vec_cap);
+            //TODO: possible micro-optimize: set iteration over ranges next_range()? getting ranges is faster in Harfbuzz int set
+            for cp in cmap_unicodes.iter() {
+                match unicode_gid_map.get(&cp) {
+                    Some(gid) => {
+                        if !input_gids.contains(*gid) && !input_unicodes.contains(cp) {
+                            continue;
+                        }
+                        self.codepoint_to_glyph.insert(cp, *gid);
+                        self.unicode_to_new_gid_list.push((cp, *gid));
+                    }
+                    None => {
+                        continue;
+                    }
+                }
+            }
+
+            /* Add gids which where requested, but not mapped in cmap */
+            //TODO: possible micro-optimize: set iteration over ranges next_range()? getting ranges is faster in Harfbuzz int set
+            for gid in input_gids.iter() {
+                if gid.to_u32() >= self.font_num_glyphs as u32 {
+                    break;
+                }
+                self.glyphset_gsub.insert(gid);
             }
         }
+        self.glyphset_gsub
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.1));
+        self.unicodes
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.0));
+    }
+
+    pub fn populate_gids_to_retain(&mut self, font: &FontRef) {
+        //not-def
+        self.glyphset_gsub.insert(GlyphId::from(0));
```

Comment:
you can use the `GlyphId::NOTDEF` constant here, which is a bit more self-documenting